### PR TITLE
Stabilize `raw_ref_op` (RFC 2582)

### LINKED ISSUE #64490: Tracking issue for RFC 2582, `&raw [mut | const] $place` (raw_ref_op)
This is a tracking issue for the RFC "an operator to take a raw reference" (rust-lang/rfcs#2582), feature(raw_ref_op).

**Steps:**

- [x] Implement the RFC (see [this comment](https://github.com/rust-lang/rust/issues/64490#issuecomment-531579111) for a detailed checklist)
- [x] Adjust documentation ([see instructions on rustc-guide][doc-guide]): https://github.com/rust-lang/reference/pull/1567
- [ ] Stabilization PR ([see instructions on rustc-guide][stabilization-guide]): https://github.com/rust-lang/rust/pull/127679

[stabilization-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#documentation-prs

**Unresolved questions:**

(None left here, since the lint was moved to https://github.com/rust-lang/rust/issues/127724)

**Implementation history:**

* Initial implementation: https://github.com/rust-lang/rust/pull/64588

### PR DESCRIPTION
**Pull request by RalfJung**
This stabilizes the syntax `&raw const $expr` and `&raw mut $expr`. It has existed unstably for ~4 years now, and has been exposed on stable via the `addr_of` and `addr_of_mut` macros since Rust 1.51 (released more than 3 years ago). I think it has become clear that these operations are here to stay. So it is about time we give them proper primitive syntax. This has two advantages over the macro:

- Being macros, `addr_of`/`addr_of_mut` could in theory do arbitrary magic with the expression on which they work. The only "magic" they actually do is using the argument as a place expression rather than as a value expression. Place expressions are already a subtle topic and poorly understood by many programmers; having this hidden behind a macro using unstable language features makes this even worse. Conversely, people do have an idea of what happens below `&`/`&mut`, so we can make the subtle topic a lot more approachable by connecting to existing intuition.
- The name `addr_of` is quite unfortunate from today's perspective, given that we have accepted provenance as a reality, which means that a pointer is *not* just an address. Strict provenance has a method, `addr`, which extracts the address of a pointer; using the term `addr` in two different ways is quite unfortunate. That's why this PR soft-deprecates `addr_of` -- we will wait a long time before actually showing any warning here, but we should start telling people that the "addr" part of this name is somewhat misleading, and `&raw` avoids that potential confusion.

In summary, this syntax improves developers' ability to conceptualize the operational semantics of Rust, while making a fundamental operation frequently used in unsafe code feel properly built in.

Possible questions to consider, based on the RFC and [this](https://github.com/rust-lang/rust/issues/64490#issuecomment-1163802912) great summary by @CAD97:

- Some questions are entirely about the semantics. The semantics are the same as with the macros so I don't think this should have any impact on this syntax PR. Still, for completeness' sake:
  - Should `&raw const *mut_ref` give a read-only pointer?
    - Tracked at: https://github.com/rust-lang/unsafe-code-guidelines/issues/257
    - I think ideally the answer is "no". Stacked Borrows says that pointer is read-only, but Tree Borrows says it is mutable.
  - What exactly does `&raw const (*ptr).field` require? Answered in [the reference](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html): the arithmetic to compute the field offset follows the rules of `ptr::offset`, making it UB if it goes out-of-bounds. Making this a safe operation (using `wrapping_offset` rules) is considered too much of a loss for alias analysis.
- Choose a different syntax? I don't want to re-litigate the RFC. The only credible alternative that has been proposed is `&raw $place` instead of `&raw const $place`, which (IIUC) could be achieved by making `raw` a contextual keyword in a new edition. The type is named `*const T`, so the explicit `const` is consistent in that regard. `&raw expr` lacks the explicit indication of immutability. However, `&raw const expr` is quite a but longer than `addr_of!(expr)`.
- Shouldn't we have a completely new, better raw pointer type instead? Yes we all want to see that happen -- but I don't think we should block stabilization on that, given that such a nicer type is not on the horizon currently and given the issues with `addr_of!` mentioned above. (If we keep the `&raw $place` syntax free for this, we could use it in the future for that new type.)
- What about the lint the RFC talked about? It hasn't been implemented yet.  Given that the problematic code is UB with or without this stabilization, I don't think the lack of the lint should block stabilization.
  - I created an issue to track adding it: https://github.com/rust-lang/rust/issues/127724
- Other points from the "future possibilites of the RFC
  - "Syntactic sugar" extension: this has not been implemented. I'd argue this is too confusing, we should stick to what the RFC suggested and if we want to do anything about such expressions, add the lint.
  - Encouraging / requiring `&raw` in situations where references are often/definitely incorrect: this has been / is being implemented. On packed fields this already is a hard error, and for `static mut` a lint suggesting raw pointers is being rolled out.
  - Lowering of casts: this has been implemented. (It's also an invisible implementation detail.)
  - `offsetof` woes: we now have native `offset_of` so this is not relevant any more.

To be done before landing:

- [x] Suppress `unused_parens` lint around `&raw {const|mut}` expressions
  - See bottom of https://github.com/rust-lang/rust/pull/127679#issuecomment-2264073752 for rationale
  - Implementation: https://github.com/rust-lang/rust/pull/128782
- [ ] Update the Reference.
  - https://github.com/rust-lang/reference/pull/1567

Fixes https://github.com/rust-lang/rust/issues/64490

cc @rust-lang/lang @rust-lang/opsem

try-job: x86_64-msvc
try-job: test-various
try-job: dist-various-1
try-job: armhf-gnu
try-job: aarch64-apple

### CODE REVIEW

**Comment by RalfJung** [comment]
@rust-lang/lang nominating for discussion. See the PR description for a summary of all relevant details. :)

Cc @rust-lang/opsem

**Comment by nikomatsakis** [comment]
I'm having mixed feelings, but it's really just... vibes I guess. Overall I've found using `addr_of!` to be "kind of ok". Obviously `&raw const a.b` is more built-in, but I'm not 100% convinced it carries its weight, and it looks "kind of funny". Though I do want to make authoring unsafe code more ergonomic in general; I don't subscribe to the philosophy that we ought to add syntactic salt just because it's unsafe, in fact I think that's quite counterproductive.

I think for me the bigger challenge is that I'm still not 100% clear on the "mental model" for validity requirements of references vs raw pointers.

**Comment by saethlin** [comment]
> I think for me the bigger challenge is that I'm still not 100% clear on the "mental model" for validity requirements of references vs raw pointers.

How is that related to this stabilization?

**Comment by RalfJung** [comment]
@nikomatsakis what about the two first points in the PR, the arguments in favor of stabilizing this?

I am happy to elaborate on the raw pointer op.sem and what is and is not UB, if you have any specific questions.

**Comment by joshtriplett** [comment]
I'm going to go ahead and propose FCP, to asynchronously see how close we are to consensus. (This does not preclude the filing of concerns.)

@rfcbot merge

**Comment by traviscross** [comment]
@rfcbot reviewed

While the `addr_of!` and `addr_of_mut!` macros were a good stopgap, raw pointers are important enough in Rust that they seem to justify having a reference-style primitive constructor syntax.

I re-read RFC 2582 and the marathon [thread](https://github.com/rust-lang/rfcs/pull/2582) for it after RalfJ posted this stabilization, and the outcome that was arrived at seemed sound and well motivated.  Every angle that came to my mind seemed to have been considered.  I'm comfortable there's not a lot more waiting to be discovered in the design space here.

**Comment by nikomatsakis** [comment]
> > I think for me the bigger challenge is that I'm still not 100% clear on the "mental model" for validity requirements of references vs raw pointers.
> 
> How is that related to this stabilization? I follow the rest of your comment, but not this last bit.

It's related only in that I'm saying -- I agree that we can do more to make raw pointers usable, but I don't think the spelling of `addr_of!` is a significant problem.

**Comment by tmandry** [comment]
I wasn't a part of the original RFC thread, but I tend to agree that a more dedicated syntax that doesn't include the word `addr` would be nice to have.

The prefix syntax often isn't great generally, but this will make raw pointers feel more integrated into, and supported by, the language. At the very least that's a subjective difference that I think is worth supporting.

@rfcbot reviewed

**Comment by tmandry** [comment]
@rfcbot concern motivation

Sorry to do this. When I reviewed this proposal I was leaning on the heuristic of "I wasn't around for the RFC and don't need to relitigate pieces like motivation".  But the RFC was quite a long time ago, and after rereading this thread it looks like @nikomatsakis and @scottmcm both have concerns about how well motivated the solution is. I don't feel convinced either. Since Niko is on vacation it's hard to know if he would block; I'm filing this concern partly on account of that, and partly because of my own reservations.

I replied inline to some points below.. but my reading now is that this feature

* Is purely syntactic – it would be _nice_, but it doesn't fundamentally solve problems that we couldn't solve in other ways,
* Doesn't significantly improve ergonomics over `addr_of!` – in fact, it's actually longer to type, and makes it harder to invoke pointer methods,
* Is more difficult to visually parse, which _might_ lead to incorrect code given that pointers allow type casting, though I could be convinced otherwise, and
* Might actually make it harder to improve pointer types later (though as the description says, there's no proposal for this on the table)

Taken together, I'm finding myself unconvinced that we should be doing this now.

> Being macros, `addr_of`/`addr_of_mut` could in theory do arbitrary magic with the expression on which they work. The only "magic" they actually do is using the argument as a place expression rather than as a value expression.

I don't find this very motivating.. sure, macros can do anything, but if you really want to know you can look at their definition.

> Place expressions are already a subtle topic and poorly understood by many programmers; having this hidden behind a macro makes this even worse.

How would this stabilization improve matters here? It seems like we are giving people new builtin magic instead of teaching them about the subtle topic. (Which might be the right thing to do – I just don't see how this being a subtle topic counts as motivation.)

> The name `addr_of` is quite unfortunate from today's perspective, given that we have accepted provenance as a reality, which means that a pointer is _not_ just an address.

I agree that the name is unfortunate given provenance, but we could literally make a new macro with a different name if that's the main issue.

> Shouldn't we have a completely new, better raw pointer type instead? Yes we all want to see that happen

I've heard references to this but don't actually know what it's referring to.

> If we keep the `&raw $place` syntax free for this, we could use it in the future for that new type.

And also `&mut raw $place`, right?

Either way, I think having two similar syntaxes that mean different things would be confusing. I don't find it compelling as a future possibility, especially when this stabilization doesn't unblock anything.

> TBH, I'm kinda torn here because I've kinda liked the grouping that the macro provides, and I end up wanting to call methods on it often if I'm using a raw pointer -- like `addr_of!(x).cast()` -- both of which are kinda elegant with the macro.

When I first heard this my response was "sure but that's a pre-existing part of `&` syntax", but I think I should've listened more closely to what @scottmcm was saying. People who need to use raw pointers don't benefit from ergonomic mitigations like auto-ref on method receivers. If you were to mitigate this yourself by leaving parens around a bare expression like `(&raw const (*foo).thing)` rustfmt will remove them, so you have to go back and add them again if you decide to call a method. While we seem to be trying to make people's lives easier here, it seems questionable whether we are.

**Comment by RalfJung** [comment]
> > Being macros, addr_of/addr_of_mut could in theory do arbitrary magic with the expression on which they work. The only "magic" they actually do is using the argument as a place expression rather than as a value expression.
> 
> I don't find this very motivating.. sure, macros can do anything, but if you really want to know you can look at their definition.

The definition uses an unstable language feature though, so you can't really look at the definition and understand what it does in the scope of stable Rust. It'd be like looking at the compiler internal implementation of an intrinsic to figure out what it does -- the result can be informative, but never authoritative.

> > Place expressions are already a subtle topic and poorly understood by many programmers; having this hidden behind a macro makes this even worse.
>
> How would this stabilization improve matters here? It seems like we are giving people new builtin magic instead of teaching them about the subtle topic. (Which might be the right thing to do – I just don't see how this being a subtle topic counts as motivation.)

It's not new magic, it's old magic. `&` and `&mut` have existed since 1.0.  By using the syntax people are already familiar with, we're cutting away a significant barrier to understanding what `addr_of!` does.

People might not understand what exactly we mean when we say "this is a place expression", but people do have an idea of what happens below `&`/`&mut`, so we can make the subtle topic a lot more approachable by connecting to existing intuition. Right now, it can easily seem like there *extra new magic* behind `addr_of!` when that is really not the case.

> I agree that the name is unfortunate given provenance, but we could literally make a new macro with a different name if that's the main issue.

When have we *ever* used a macro as the permanent and only stable way to expose a language feature? I feel like this kind of proposal needs some very serious justification.

We use macros for experimentation (`try!`, `addr_of!`), and then if the feature carries its weight we give it proper syntax.

In this case, the macro also has the big downside of entirely removing the symmetry with `&` and `&mut`, thus imposing an extra burden to understanding what this operation does.

> Either way, I think having two similar syntaxes that mean different things would be confusing. I don't find it compelling as a future possibility, especially when this stabilization doesn't unblock anything.

Having `&` and `&mut` mean different things isn't confusing, so I don't understand how this would be any more confusing in a "new pointer type world". After all, these are different *but very related*, parallel things. Parallel concepts *should* have similar syntax.

Also, indefinitely blocking proper support for our existing raw pointer types on the completely hypothetical idea for a shiny new raw pointer type *which hasn't seen any tangible process in the many years since it has been first discussed* doesn't seem like a great idea to me. It is frustrating to see attempts at small local improvements blocked on such very hypothetical ideas.

>  If you were to mitigate this yourself by leaving parens around a bare expression like (&raw const (*foo).thing) rustfmt will remove them

This is something that can be changed.

**Comment by tmandry** [comment]
> > > Being macros, addr_of/addr_of_mut could in theory do arbitrary magic with the expression on which they work. The only "magic" they actually do is using the argument as a place expression rather than as a value expression.
> > 
> > 
> > I don't find this very motivating.. sure, macros can do anything, but if you really want to know you can look at their definition.
> 
> The definition uses an unstable language feature though, so you can't really look at the definition and understand what it does in the scope of stable Rust. It'd be like looking at the compiler internal implementation of an intrinsic to figure out what it does -- the result can be informative, but never authoritative.
> 
> > > Place expressions are already a subtle topic and poorly understood by many programmers; having this hidden behind a macro makes this even worse.
> > 
> > 
> > How would this stabilization improve matters here? It seems like we are giving people new builtin magic instead of teaching them about the subtle topic. (Which might be the right thing to do – I just don't see how this being a subtle topic counts as motivation.)
> 
> It's not new magic, it's old magic. `&` and `&mut` have existed since 1.0. By using the syntax people are already familiar with, we're cutting away a significant barrier to understanding what `addr_of!` does.
> 
> People might not understand what exactly we mean when we say "this is a place expression", but people do have an idea of what happens below `&`/`&mut`, so we can make the subtle topic a lot more approachable by connecting to existing intuition. Right now, it can easily seem like there _extra new magic_ behind `addr_of!` when that is really not the case.

Thank you, this is much clearer to me in terms of motivation. Improving developers' ability to conceptualize the operational semantics of Rust does seem important, especially for the correctness and soundness of unsafe code.

I'll check that I'm not forgetting anything, but I think this resolves most or all of my concern.

> > If you were to mitigate this yourself by leaving parens around a bare expression like (&raw const (*foo).thing) rustfmt will remove them
> 
> This is something that can be changed.

Yes, I suppose it can, and maybe it should be.

While I do want to give the ergonomic question due consideration, I feel optimistic. At the very worst we could go ahead with stabilization and choose not to (soft-)deprecate `addr_of!()` until we had a solution that felt roughly as ergonomic as before. Though I'd much prefer to do it all at once, so people have clear guidance.

Separately from this proposal I'm interested in the idea of postfix `.&`, `.&mut`, `.*`, and after this, `.&raw`. That would also solve the method dispatch issue.

> When have we _ever_ used a macro as the permanent and only stable way to expose a language feature? I feel like this kind of proposal needs some very serious justification.

I wasn't actually proposing that. I was saying that if the name is the biggest issue we can address that directly, until we are ready to stabilize permanent syntax.

> Also, indefinitely blocking proper support for our existing raw pointer types on the completely hypothetical idea for a shiny new raw pointer type _which hasn't seen any tangible progress or even serious attempts at progress in the many years since it has been first discussed_ doesn't seem like a great idea to me. It is honestly quite frustrating to see attempts at small local improvements blocked on such very hypothetical ideas.

I agree that we shouldn't block on such things. My last comment is asking whether this is a real improvement on any axis of importance. Your response makes me think that it is.

To clarify where I'm coming from, I don't think we should hold out on syntax for distant hypotheticals. The general situation I am trying to avoid is using up syntax for a solution that we know has problems if we're only doing it because providing a dedicated syntax feels like the "right thing to do".

> Having `&` and `&mut` mean different things isn't confusing, so I don't understand how this would be any more confusing in a "new pointer type world". After all, these are different _but very related_, parallel things. Parallel concepts _should_ have similar syntax.

I think we might be talking about different things? Not sure. I'm saying that if in the future we had `&raw mut x` mean one type and `&mut raw x` mean another it would be highly confusing. Remembering which order to use for which case would be a real burden. We could mitigate that by deprecating the old syntax over an edition, but it's not a prospect that brings me joy.

**Comment by RalfJung** [comment]
Thanks for clarifying, this helps!

> I think we might be talking about different things? Not sure. I'm saying that if in the future we had &raw mut x mean one type and &mut raw x mean another it would be highly confusing. Remembering which order to use for which case would be a real burden.

Ah yes, agreed.
IIRC the "future better raw pointer type" is often considered to only come in one variant, not two (`mut` and `const`). So there might not be a need for a `mut` variant of it. But then, it's hard to say what it will look like.

> At the very worst we could go ahead with stabilization and choose not to (soft-)deprecate addr_of!() until we had a solution that felt roughly as ergonomic as before.

I'm not strongly attached to the soft-deprecation.

I don't know when exactly rustfmt currently auto-removes parentheses, and in which cases you'd prefer that to not happen. Are you thinking of `let ptr = (&raw const <expr>);`?

**Comment by CAD97** [comment]
> IIRC the "future better raw pointer type" is often considered to only come in one variant, not two (`mut` and `const`).

The difficulty there is that such a type would need to be *bi*variant in order to support usage in both *co*variant and *contra*variant types, and bivariance basically means unconstrained to the rustc type system[^1]. An opaque pointer type without a pointee type is useful; one that carries a pointee type but is variance-unified with any other pointee type seems problematic.

[^1]: It's [possible to construct](https://github.com/lcnr/random-rust-snippets/issues/3) a bivariant type parameter despite unused type parameter errors, by utilizing associated type protection.

----------

My primary hesitation to stabilize the `&raw mut $place` has been the thought that we might use `&raw` as the type syntax for the improved raw pointer type, and want the expression to make that pointer type. That type might've been as simple as `*mut T` but non-null, or as involved as `MaybeDangling<unsafe<'a> &'a mut T`>, or just `ptr::WellFormed<T>`. But I don't think it's worth blocking the syntax for use today, for two main reasons: 

- An improved raw pointer type is still highly aspirational, and nowhere near being reality. 
- It's reasonably possible to make the new type decay to today's pointer type, and transition the syntax to produce the new type over an edition, if we ever find a solution to defaults affecting inference sets that we so desperately desire.

So consider this the equivalent of a checkbox from my end — this is an improvement, and it does not meaningfully prohibit desired future improvements.

**Comment by RalfJung** [comment]
> The difficulty there is that such a type would need to be bivariant in order to support usage in both covariant and contravariant types,

I don't see why it would need to support contravariant types, that's an extremely rare circumstance and already not supported with our current raw pointer types.

I also don't think it should be the job of a new raw pointer type to fix the problems with our variance system; that should instead be done by allowing explicit variance annotations for type parameters of a generic type. The raw pointer type can then be invariant for a maximally safe default, and types that want covariance (or contravariance) can opt-in to that explicitly.

But anyway this is getting rather off-topic for this thread.

**Comment by tmandry** [comment]
@rfcbot resolve motivation

This stabilization has a big impact in terms of syntax, but it makes it easier for Rust users to learn how to use unsafe. The negative ergonomic impacts of moving from `addr_of!` can be mitigated in the ways I mention below.

> I don't know when exactly rustfmt currently auto-removes parentheses, and in which cases you'd prefer that to not happen. Are you thinking of `let ptr = (&raw const <expr>);`?

Yes. Actually I misremembered; it's the rustc `unnecessary_parentheses` lint, not rustfmt that enforces this.

Maybe if we just relax that lint in the case where it's around `&raw {const|mut}` expressions, that's good enough. I would be comfortable telling people to use this instead of `addr_of!` then. It's not as good as `.&raw` but that can come later.

> * Is more difficult to visually parse, which _might_ lead to incorrect code given that pointers allow type casting, though I could be convinced otherwise, and

Following up on my own point, I can't think of examples where this would lead to bugs.

**Comment by tmandry** [comment]
@rfcbot concern mitigate the ergonomic impact or don't soft-deprecate addr_of

**Comment by RalfJung** [comment]
> mitigate the ergonomic impact or don't soft-deprecate addr_of

Does "suppress `unnecessary_parentheses`" count as "mitigate the ergonomic impact"?

**Comment by tmandry** [comment]
> Does "suppress `unnecessary_parentheses`" count as "mitigate the ergonomic impact"?
> 
> We can add that to the FCP then, making it required before landing the PR.

Yes. I took the liberty of updating the PR description with this (and elaborated on the rationale based on your recent comments). Let's move it forward.

@rfcbot resolve mitigate the ergonomic impact or don't soft-deprecate addr_of

**Comment by RalfJung** [comment]
The lint change is happening in https://github.com/rust-lang/rust/pull/128782, and I view the FCP here as also gathering approval for this lint change.

**Comment by ehuss** [comment]
Before this is merged, can we please have a reference documentation PR posted and reviewed?

**Comment by RalfJung** [comment]
Inside the compiler this was still called the "address-of" operator, which I figured should also be fixed. So I renamed it to "raw borrow" operator. If someone has a better idea for what to call the operator, please speak up. :)

**Comment by traviscross** [comment]
r? compiler

This is still in FCP, but it's likely FCP will complete soon, so this is worth a compiler review at this point.

**Review by traviscross — APPROVED** [review]
The code changes look right to me.

**Comment by jieyouxu** [comment]
I'll take a look a this ~today/tomorrow, I'll need to read the RFC and context a bit.

File: tests/ui/raw-ref-op/feature-raw-ref-op.rs
Comment (jieyouxu) [inline tests/ui/raw-ref-op/feature-raw-ref-op.rs:1]: Remark: I almost thought this was a compiletest directive I missed while migrating. It might just be a copy-pasta from another test?

File: library/core/src/ptr/mod.rs
Comment (jieyouxu) [inline library/core/src/ptr/mod.rs:2213]: Remark (not in scope of this PR): this probably wants an allow-by-default deprecation lint or something eventually that can be bumped up to a warning.

**Review by jieyouxu — APPROVED** [review]
Thanks, the changes LGTM. I'll run a couple of try jobs on different platforms just in case there's `cfg`'d out code or ignored tests that were missed.

**Comment by jieyouxu** [comment]
Also minor nit: the PR description saying "fix ..." would automatically cause the tracking issue #64490 to be closed when this PR is merged, but the tracking issue should probably be closed after the reference PR is merged, which is blocked by this PR.

**Comment by jieyouxu** [comment]
Welp, that's my fault for re-enable it 😆

File: tests/ui/raw-ref-op/feature-raw-ref-op.rs
Comment (RalfJung) [reply tests/ui/raw-ref-op/feature-raw-ref-op.rs:1]: No, `gate-test` exists in a bunch of places. It is interpreted by tidy, not compiletest, which is why it doesn't have the `@`.

File: library/core/src/ptr/mod.rs
Comment (RalfJung) [reply library/core/src/ptr/mod.rs:2213]: We have "deprecated in a future version of Rust" that we can use for that, but I intended for this to be even softer to avoid too much churn.

**Comment by RalfJung** [comment]
May I ask what is the purpose of these try builds?

> Also minor nit: the PR description saying "fix ..." would automatically cause the tracking issue https://github.com/rust-lang/rust/issues/64490 to be closed when this PR is merged, but the tracking issue should probably be closed after the reference PR is merged, which is blocked by this PR.

Not sure in which order we usually do this -- should the implementation PR close the tracking issue, or should the reference PR do that?

> (The tracking issue also has an open unresolved question, I'm not sure what's the status with that.)

That's about the lint, so it should be tracked in https://github.com/rust-lang/rust/issues/127724 now.

File: tests/ui/raw-ref-op/feature-raw-ref-op.rs
Comment (jieyouxu) [reply tests/ui/raw-ref-op/feature-raw-ref-op.rs:1]: *oh* thanks, i didnt consider tidy

**Comment by jieyouxu** [comment]
> May I ask what is the purpose of these try builds?

Just sanity checks because I'm paranoid due to reviewing run-make tests. I don't think you missed any tests.

Feel free to r=me and TC after the try-jobs come back green.

**Comment by RalfJung** [comment]
I don't think this PR is particularly likely to cause strange problems that wouldn't be caught by PR CI, and even then if we don't have any concrete suspicion generally I think we just `r+` these and let bors catch those cases. The queue is also rather empty right now so it's not like we'd be wasting previous bors slots.

**Comment by jieyouxu** [comment]
Fair enough. Unfortunately I can't cancel the try-jobs now, and we can't r+ immediately either when a try build is on-going or else bors can (or will) push this to master without any testing.

**Comment by jieyouxu** [comment]
> Not sure in which order we usually do this -- should the implementation PR close the tracking issue, or should the reference PR do that?

For tracking purposes, I think it's most logical if whichever merges last closes the tracking issue.

**Comment by RalfJung** [comment]
Okay, let's see if cross-repo issue closing with a reference PR works. :)

**Comment by jieyouxu** [comment]
Thanks!
@bors r=traviscross,jieyouxu rollup

**Comment by RalfJung** [comment]
Rebasing succeeded without any conflicts... :shrug: 

@bors r=jieyouxu

**Comment by jieyouxu** [comment]
... does those test need blessing, or is this comment outdated I wonder

**Comment by RalfJung** [comment]
Probably just needs blessing.

@bors r-

**Pull request merged**
